packages/x-date-pickers/src/internals/hooks/useField/useField.ts

```diff
@@ -130,6 +141,128 @@ export const useField = <
     updateSelectedSections();
   });
 
+  const handleInputChange = useEventCallback((event: React.ChangeEvent<HTMLInputElement>) => {
+    if (readOnly || state.selectedSectionIndexes == null) {
+      return;
+    }
+
+    const prevValueStr = fieldValueManager.getValueStrFromSections(state.sections);
+    const valueStr = event.target.value;
+
+    let startOfDiffIndex = -1;
+    let endOfDiffIndex = -1;
+    for (let i = 0; i < prevValueStr.length; i += 1) {
+      if (startOfDiffIndex === -1 && prevValueStr[i] !== valueStr[i]) {
+        startOfDiffIndex = i;
+      }
+
+      if (
+        endOfDiffIndex === -1 &&
+        prevValueStr[prevValueStr.length - i - 1] !== valueStr[valueStr.length - i - 1]
+      ) {
+        endOfDiffIndex = i;
+      }
+    }
```

Comment:
I'm not super confident with this method. I'm surprised it does not bugg when a character of the previous value is included in the new one.

For example if it was 12 and you copy-past 22, this algo will detect that a 1 has been added.

I would feel safer by using `getValueStrFromSections` to generate `beforeSections` and `afterSections` which would be the string generated before the active section and after the active section.

For that you can get the active section: with `valueStr.slice(beforeSections.lenght, valueStr.length - afterSection.length)`

And you can check their is not modification outside with `valueStr.slice(0, beforeSections.lenght) === beforeSections`

Reply:
> I would feel safer by using getValueStrFromSections to generate beforeSections and afterSections which would be the string generated before the active section and after the active section.

I think it does not work on Android since on the 2nd call the value is the value with the selected section cleaned but the sections in state is not cleaned.
Maybe we can keep the section up to date with this weird behavior.

I'd prefer to merge #6141 which touches a lot those behaviors.